gdrdrv/gdrdrv.c

```diff
@@ -378,6 +433,39 @@ static void gdrdrv_get_pages_free_callback(void *data)
 
 //-----------------------------------------------------------------------------
 
+/**
+ * Generate mr->handle. This function should be called under info->lock.
+ *
+ * Prerequisite:
+ * - mr->mapped_size is set and round to max(PAGE_SIZE, GPU_PAGE_SIZE)
+ *
+ * Return 0 if success, -1 if failed.
+ */
+static inline int gdr_generate_mr_handle(gdr_info_t *info, gdr_mr_t *mr)
+{
+    // The user-space library passes the memory (handle << PAGE_SHIFT) as the
+    // mmap offset, and offsets are used to determine the VMAs to delete during
+    // invalidation.  
+    // Hence, we need [(handle << PAGE_SHIFT), (handle << PAGE_SHIFT) + size - 1] 
+    // to correspond to a unique VMA.  Note that size here must match the
+    // original mmap size
+
+    gdr_hnd_t next_handle = info->next_handle + (mr->mapped_size >> PAGE_SHIFT);
+
+    // The end of this range is wrapped around. The calculation uses PAGE_SHIFT
+    // because handle is equal vm_pgoff (aka file offset << PAGE_SHIFT). We
+    // indicate this case as -ENOMEM error.
+    if ((next_handle & ((gdr_hnd_t)(-1) >> PAGE_SHIFT)) < info->next_handle)
```

Comment:
I don't like failing eagerly. It should fail when next handle is going to be used.